File: fm-domain/src/main/java/io/zeitwert/fm/dms/adapter/api/jsonapi/impl/ObjDocumentDtoAdapter.java
package io.zeitwert.fm.dms.adapter.api.jsonapi.impl;

import io.zeitwert.dddrive.ddd.adapter.api.jsonapi.dto.EnumeratedDto;
import io.zeitwert.fm.dms.adapter.api.jsonapi.dto.ObjDocumentDto;
import io.zeitwert.fm.dms.model.ObjDocument;
import io.zeitwert.fm.dms.model.enums.CodeContentKind;
import io.zeitwert.fm.dms.model.enums.CodeDocumentCategory;
import io.zeitwert.fm.dms.model.enums.CodeDocumentKind;
import io.zeitwert.fm.obj.adapter.api.jsonapi.base.ObjDtoAdapterBase;
import org.springframework.stereotype.Component;

@Component("objDocumentDtoAdapter")
public class ObjDocumentDtoAdapter extends ObjDtoAdapterBase<ObjDocument, ObjDocumentDto> {

	@Override
	public void toAggregate(ObjDocumentDto dto, ObjDocument obj) {
		try {
			obj.getMeta().disableCalc();
			super.toAggregate(dto, obj);
			obj.setName(dto.getName());
			obj.setContentKind(dto.getContentKind() == null ? null
					: CodeContentKind.getContentKind(
					dto.getContentKind().getId()));
			obj.setDocumentKind(dto.getDocumentKind() == null ? null : CodeDocumentKind.getDocumentKind(dto.getDocumentKind().getId()));
			obj.setDocumentCategory(dto.getDocumentCategory() == null ? null
					: CodeDocumentCategory.getDocumentCategory(
					dto.getDocumentCategory().getId()));
		} finally {
			obj.getMeta().enableCalc();
			obj.getMeta().calcAll();
		}
	}

	@Override
	public ObjDocumentDto fromAggregate(ObjDocument obj) {
		if (obj == null) {
			return null;
		}
		ObjDocumentDto.ObjDocumentDtoBuilder<?, ?> dtoBuilder = ObjDocumentDto.builder();
		this.fromAggregate(dtoBuilder, obj);
		// @formatter:off
		return dtoBuilder
			.name(obj.getName())
			.contentKind(EnumeratedDto.of(obj.getContentKind()))
			.supportedContentTypes(obj.getContentKind().getExtensions().stream().reduce("", (a, b) -> a.length() > 0 ? a + "," + b : b))
			.documentKind(EnumeratedDto.of(obj.getDocumentKind()))
			.documentCategory(EnumeratedDto.of(obj.getDocumentCategory()))
			.contentType(EnumeratedDto.of(obj.getContentType()))
			.build();
		// @formatter:on
	}

//	@Override
//	public ObjDocumentDto fromRecord(ObjDocumentVRecord obj) {
//		if (obj == null) {
//			return null;
//		}
//		ObjDocumentDto.ObjDocumentDtoBuilder<?, ?> dtoBuilder = ObjDocumentDto.builder();
//		this.fromRecord(dtoBuilder, obj);
//		// @formatter:off
//		return dtoBuilder
//			.name(obj.getName())
//			.contentKind(EnumeratedDto.of(CodeContentKindEnum.getContentKind(obj.getContentKindId())))
//			.documentKind(EnumeratedDto.of(CodeDocumentKindEnum.getDocumentKind(obj.getDocumentKindId())))
//			.documentCategory(EnumeratedDto.of(CodeDocumentCategoryEnum.getDocumentCategory(obj.getDocumentCategoryId())))
//			.build();
//		// @formatter:on
//	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/dms/adapter/jsonapi/impl/ObjDocumentDtoAdapter.kt
package io.zeitwert.fm.dms.adapter.jsonapi.impl

import dddrive.ddd.model.RepositoryDirectory
import io.zeitwert.app.api.jsonapi.dto.EnumeratedDto
import io.zeitwert.app.obj.api.jsonapi.base.ObjDtoAdapterBase
import io.zeitwert.fm.dms.adapter.jsonapi.dto.ObjDocumentDto
import io.zeitwert.fm.dms.model.ObjDocument
import org.springframework.stereotype.Component

@Component("objDocumentDtoAdapter")
class ObjDocumentDtoAdapter(
	directory: RepositoryDirectory,
) : ObjDtoAdapterBase<ObjDocument, ObjDocumentDto>(
		ObjDocument::class.java,
		"document",
		ObjDocumentDto::class.java,
		directory,
		{ ObjDocumentDto() },
	) {

	init {
		config.field(
			"contentType",
			outgoing = { EnumeratedDto.of((it as ObjDocument).contentType) },
		)

		// supportedContentTypes is a calculated field from contentKind.getExtensions()
		config.field(
			"supportedContentTypes",
			outgoing = { (it as ObjDocument).contentKind?.getExtensions()?.joinToString(",") ?: "" },
		)
	}

}
